plugins/node/opentelemetry-instrumentation-restify/src/instrumentation.ts

```diff
@@ -0,0 +1,204 @@
+ * limitations under the License.
+ */
+
+import * as api from '@opentelemetry/api';
+import * as restify from 'restify';
+import { Server } from 'restify';
+import * as types from './types';
+import { VERSION } from './version';
+import once = require('lodash.once');
+import {
+  InstrumentationBase,
+  InstrumentationNodeModuleDefinition,
+  InstrumentationNodeModuleFile,
+  isWrapped,
+} from '@opentelemetry/instrumentation';
+import { HttpAttribute } from '@opentelemetry/semantic-conventions';
+
+const { diag } = api;
+const RESTIFY_MW_METHODS = ['use', 'pre'];
+const RESTIFY_METHODS = ['del', 'get', 'head', 'opts', 'post', 'put', 'patch'];
+
+const MODULE_NAME = 'restify';
+const SUPPORTED_VERSIONS = ['>=4.0.0'];
+
+export class RestifyInstrumentation extends InstrumentationBase<
+  typeof restify
+> {
+  constructor() {
+    super(`@opentelemetry/instrumentation-${MODULE_NAME}`, VERSION);
+  }
+
+  private _moduleVersion?: string;
+  private _isDisabled = false;
+
+  init() {
+    const module = new InstrumentationNodeModuleDefinition<typeof restify>(
+      MODULE_NAME,
+      SUPPORTED_VERSIONS,
+      (moduleExports, moduleVersion) => {
+        this._moduleVersion = moduleVersion;
+        return moduleExports;
+      }
+    );
+
+    module.files.push(
+      new InstrumentationNodeModuleFile<typeof restify>(
+        'restify/lib/server.js',
+        SUPPORTED_VERSIONS,
+        (moduleExports, moduleVersion) => {
+          diag.debug(`Applying patch for ${MODULE_NAME}@${moduleVersion}`);
+          this._isDisabled = false;
+          const Server: any = moduleExports;
+          for (const name of RESTIFY_METHODS) {
+            if (isWrapped(Server.prototype[name])) {
+              this._unwrap(Server.prototype, name);
+            }
+            this._wrap(
+              Server.prototype,
+              name as keyof Server,
+              this._methodPatcher.bind(this)
+            );
+          }
+          for (const name of RESTIFY_MW_METHODS) {
+            if (isWrapped(Server.prototype[name])) {
+              this._unwrap(Server.prototype, name);
+            }
+            this._wrap(
+              Server.prototype,
+              name as keyof Server,
+              this._middlewarePatcher.bind(this)
+            );
+          }
+          return moduleExports;
+        },
+        (moduleExports, moduleVersion) => {
+          diag.debug(`Removing patch for ${MODULE_NAME}@${moduleVersion}`);
+          this._isDisabled = true;
+          if (moduleExports) {
+            const Server: any = moduleExports;
+            for (const name of RESTIFY_METHODS) {
+              this._unwrap(Server.prototype, name as keyof Server);
+            }
+            for (const name of RESTIFY_MW_METHODS) {
+              this._unwrap(Server.prototype, name as keyof Server);
+            }
+          }
+        }
+      )
+    );
+
+    return module;
+  }
+
+  private _middlewarePatcher(original: Function, methodName?: string) {
```

Comment:
I think middleware spans should have middleware name as the span name (not sure if possibly with restify), e.g. the operation tree looks like this currently, where each middleware operation name is the route name:
![image](https://user-images.githubusercontent.com/5329631/114398762-d2e58400-9ba8-11eb-9174-8fb16e56b37a.png)

However the express instrumentation contains middleware names:
![image](https://user-images.githubusercontent.com/5329631/114398846-ef81bc00-9ba8-11eb-94b5-6a9e3489ddc7.png)

Also I think middleware spans should have the incoming request span as its parent.

Reply:
Yep, it's possible. Express takes the name of the function it seems, which can be undefined at times, but still better than nothing. Thanks!

> Also I think middleware spans should have the incoming request span as its parent.

I generally agree but don't want to change that because there might not be a request span if HTTP instrumentation is not enabled, so the structure would be different between HTTP instrumentation enabled and not.